### PR TITLE
Implement policies for committers and pull requests.

### PR DESCRIPTION
**Pull request by jmchilton**
- State and clarify my (@jmchilton's) understanding of existing informal pull request policies.
- Add a list of committers who should have "commit" access to the repository according to my understanding of the team meeting held on 5/28/2015.
- Add language describing how the Galaxy project might manage commit access and project governance going forward in accordance with my understanding of the team meeting discussion.

This fills in a lot of gaps left by informal policies and informal discussions and I am eager to see much discussion on each of these topics. Much of the language and structure is borrowed with heavy modification from the Django and Apache variants of these documents.

Also - congrats to @nsoranzo, @erasche, and @bgruening who the team unanimously voted to grant commit access to.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -103,3 +102,59 @@ developers.
...
+
+## Handling Pull Requests
+
+Everyone is encouraged to express opinions and issue non-binding votes on pull
+requests, but only members of the *committers* group may issue binding votes
+on pull requests. Information on the *committers* group can be found in the
+[organization document](https://github.com/galaxyproject/galaxy/blob/dev/doc/source/project/organization.rst)
+describing governance of the core Galaxy code base.
+
+Votes should on pull requests should take the form of
```
Comment (hexylena) [inline CONTRIBUTING.md]: Maybe "Votes on pull requests should"?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -103,3 +102,59 @@ developers.
...
+exposure, negatively affects performance, etc. ). A veto without a
+justification is invalid and has no weight."
+
+### Direct Commit Access
+
+The Galaxy *committers* group may only commit directly to Galaxy (i.e.  outside
+of a pull request and not following the procedures described here) the
+following two categories of patches:
+
+* Serious security vulnerabilities must simultaneously be made to multiple
```
Comment (hexylena) [inline CONTRIBUTING.md]: Maybe "Serious security vulnerability patches, which must" (or some variant which clearly delineates the stipulation clause on "srs sec vulns")

**Comment by martenson** [comment]
:+1:

File: doc/source/project/organization.rst [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+==================================
+
+This is not a legal document and the Galaxy Project is not a legal
+entity. Pennsylvania State University owns copyright to the Galaxy Project and
+licenses the project under the terms of the Academic Free License version 3.0
+(see the LICENSE_ for full details).
+
+This document informally outlines the organizational structure governing the
+Galaxy core code base hosted at https://github.com/galaxyproject/galaxy. This
+governance extends to code related activities of this repository such as
```
Comment (nsoranzo) [inline doc/source/project/organization.rst]: s/code related/code-related/

File: doc/source/project/organization.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+advocates with commit access to the repository as of May 2015. This group
+subsequently voted in new members.
+
+Any member of the `committers` group may nominate an individual for membership
+to the `committers` group. Such individuals must have demonstrated:
+
+- a good grasp of the design of Galaxy core project
+- a solid track record of being constructive and helpful
+- significant contributions to the project
+- willingness to dedicate some time to improving Galaxy
```
Comment (nsoranzo) [inline doc/source/project/organization.rst]: Missing closing period.

Comment (martenson) [inline doc/source/project/organization.rst]: I think all the bullet points are not considered sentences in this case.

File: CONTRIBUTING.md (hunk 2 above)
Comment (martenson) [inline CONTRIBUTING.md]: I would drop the simultaneous & multiple branches part entirely.

**Comment by nsoranzo** [comment]
Many thanks to John and the Galaxy Team for trusting me so much!

**Comment by hexylena** [comment]
Likewise! I'm proud to give my first binding :+1: to this. Thank you all.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -103,3 +102,59 @@ developers.
...
+Pull requests modifying pre-existing releases should be restricted to bug fixes
+and require at least 2 *+1* binding votes from someone other than the author of
+the pull request with no *-1* binding votes.
+
+Any pull request changing or clarifying the procedures governing this
+repository must remain open for at least 4 days and require at least 3 *+1*
+votes with no *-1* votes from qualified *committers*.
+
+Any other pull request requires at least 1 binding *+1* vote from someone other
+than the author of the pull request.
```
Comment (nsoranzo) [inline CONTRIBUTING.md:139]: "someone other than the author" but still in the _committers_ group?

Comment (bgruening) [inline CONTRIBUTING.md:139]: I think so ... `binding votes` are defined as from `members of the *committers* group` above.

**Comment by bgruening** [comment]
This is a big step forward for the entire Galaxy community - Thank you all for making Galaxy even more open!

File: doc/source/project/organization.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
+==================================
+Galaxy Core Governance
+==================================
+
+This is not a legal document and the Galaxy Project is not a legal
+entity. Pennsylvania State University owns copyright to the Galaxy Project and
+licenses the project under the terms of the Academic Free License version 3.0
+(see the LICENSE_ for full details).
```
Comment (jxtx) [inline doc/source/project/organization.rst]: I think we can remove this paragraph entirely.

File: CONTRIBUTING.md (hunk 5 above)
Comment (jmchilton) [inline CONTRIBUTING.md:139]: Exactly @bgruening.

**Comment by jmchilton** [comment]
Pushed out two small adjustments corresponding to changes requested by @martenson / @erasche and @jxtx respectively.

**Comment by jxtx** [comment]
With the changes as of 0fe60d49ba3cbf2dc54cb7b045a17f732268c811, I am +1

File: CONTRIBUTING.md (hunk 5 above)
Comment (nsoranzo) [inline CONTRIBUTING.md:139]: I think I missed it because here 'binding' is before the '+1': `s/at least 1 binding *+1* vote/at least 1 *+1* binding vote/` ?

**Comment by dannon** [comment]
+1, looks great now.

**Comment by jmchilton** [comment]
I tried to approach the commit message as a moderately disinterested party clarifying the current party and filling in gaps. And while I am a strong :+1: (I have been fighting for much of this longer than I have been on the team), I did want to publicly state my reservations (which I have expressed privately). In particular I think allowing any member of the _committers_ group to have a veto on any commit (even bug fixes) is problematic. IMO - there should be a mechanism whereby a veto can be countered with some number of +1s (say 3) and that results in some sort of voting process of all _committers_ - and 4/5th or 3/5th or 5/7th or some percent like that of voting members voting +1 should allow an override. I trust the consensus of the people in the _committers_ group much more than I trust most of the individual members. And while I suspect I will be on the receiving end of -1 vetos much more than I will be on the vetoing end, I think even vetoers would be protected by voting. Their actions would be more readily interpreted as the process or the group than as personal actions as the result of such votes.

File: CONTRIBUTING.md (hunk 5 above)
Comment (jmchilton) [inline CONTRIBUTING.md:139]: Adjusted in 3316b25 - thanks for the suggestion @nsoranzo.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -103,3 +102,58 @@ developers.
...
+Votes on pull requests should take the form of
+[+1, 0, -1, and fractions](http://www.apache.org/foundation/voting.html)
+as outlined by the Apache Foundation.
+
+Pull requests modifying pre-existing releases should be restricted to bug fixes
+and require at least 2 *+1* binding votes from someone other than the author of
+the pull request with no *-1* binding votes.
+
+Any pull request changing or clarifying the procedures governing this
+repository must remain open for at least 4 days and require at least 3 *+1*
```
Comment (blankenberg) [inline CONTRIBUTING.md]: Excluding weekends/holidays? In general, I think 4 days is too short of a comment window for modifying rules of procedures.

Also, we should perhaps create a special tag for this, and only allow changes to these rules as part of an independent Pull request to prevent any changes from being masked by e.g. a larger pull request.

Comment (jmchilton) [inline CONTRIBUTING.md]: I do like the idea of the special tag and the language regarding forcing governance changes to be made in their own independent pull request.

I am -0.5 excluding holidays - too US specific. My counter proposal would be to keep 4 days but exclude weekends - is that enough - does anyone else think 4 days is insufficient?

Comment (blankenberg) [inline CONTRIBUTING.md]: I would perhaps remove consideration of weekends/holidays and instead make it at least 8 days (week +1 day), specifically stated in hours ( = 192 hours).

I would also say 3 +1 is far too few, as that is < 16% of committers. Something like ceil( max( 3, 51% of committers) ) voting +1, with no -1; although 51% (currently 10 votes) may still be too small.

IMHO, changing formalized well-established procedures is something that should happen exceptionally rarely, and only when overwhelming cause dictates.

Comment (jmchilton) [inline CONTRIBUTING.md]: @blankenberg We had everyone in a room voting and I don't think we had 10 truly positive votes for the uncontroversial rule changes (e.g. adding additional committers). I would argue we need agile governance for all the same reasons we should have agile code - so that we can adapt to changing situations.

Lazy consensus should win the day especially given the extraordinary power we are given each committer with the insurmountable -1 veto.

Comment (dannon) [inline CONTRIBUTING.md]: I'm +1 to @blankenberg's suggestions.  What sort of changing situations are you thinking we'll need to adapt to that we can't take the time to organize folks into acknowledging with a vote?

Comment (jmchilton) [inline CONTRIBUTING.md]: @dannon Like I mentioned - I don't think we got 10 +1 for any vote yet when it was a simple as raising your hand and gathering votes for these things will be much harder on github.

If I understood the ways in which this was inadequate or how situations will change - I think we would just modify it now. The fact is these procedures may be wholly terrible - we really don't know. Finally, I would wait for the current rule change procedures to actually fail us in someway before limiting our ability to modify them.

How about 4 +1s and the 192 hours?

Comment (jxtx) [inline CONTRIBUTING.md]: I agree with @jmchilton that we should not bind ourselves before we have a chance to see how this process works. Down the road if we collectively choose to change to 10 +1s so be it. 

> changing formalized well-established procedures is something that should happen exceptionally rarely, and only when overwhelming cause dictates

Totally agree, but we don't have "well-established procedures" yet in this case.

Comment (dannon) [inline CONTRIBUTING.md]: Alright. I'm ok with going with 4 (maybe 5?) +1's and 192 hours.  I don't think that's anywhere near being an unreasonable number of folks to wrangle into a conversation about changing the project procedures.

Comment (blankenberg) [inline CONTRIBUTING.md]: How about: 25% of committers (5 currently) and 192 hours for open voting, with a -1 being a veto that is unable to be overruled nor overridden. No modifications of PRs; if changes are needed, close PR and open a new one (old votes can't count after modifications). These PRs must be independent and specially tagged with [SOME_TEXT_TO_BE_DETERMINED] prepended to title and github tag at time of creation.

Comment (jmchilton) [inline CONTRIBUTING.md]: @blankenberg Okay I have grudgingly made the requested clarifications. If everyone is still on-board tomorrow - I will close this pull request and re-open - since I guess the pull request establishing the rules should follow the rules it establishes.

**Comment by natefoo** [comment]
I am :+1: on this. Should we follow the new procedure for this? A clean PR and re-vote?

**Comment by jmchilton** [comment]
@natefoo Yup - I'll give everyone a few more work hours to review things and reopen this tomorrow afternoon.

**Comment by nsoranzo** [comment]
:+1: to the procedure refinement.

**Comment by natefoo** [comment]
I think it's maybe worth noting that @jmchilton's point about vetoes has another implication because of the way our procedure change system is defined. Let's say someone is -1ing code PRs and the rest of the committers decide that a process change is needed to allow for a veto override (that 4/5 or whatever). Because a -1 in process changes can also stop a change dead, there is now no way to change the policy if that chronic -1er also -1s the policy change PR.

Not that I think anyone is going to be that -1er (and there are potentially other incentives for some of us to play nicely since there is the issue of employment...), but a lack of veto override does potentially back us into a corner that we could, theoretically, not recover from without breaking our procedure.

**Comment by hexylena** [comment]
@natefoo,

> and there are potentially other incentives for some of us to play nicely since there is the issue of employment...

it might be worth considering that now that there will be external committers, not all of us will be under that threat, so it might really be worth considering veto override procedures.

**Comment by natefoo** [comment]
@erasche, @jmchilton pointed out that the committer removal policy gives us a way to deal with it, since people cannot vote on their own removal vote.

> Pull requests modifying the membership of this list are subject to the normal rules for pull requests that modify governance procedures outlined in Galaxy's CONTRIBUTING document - with one exception - a `commmitter` may not vote against their own removal from the group (for obvious reasons).

That said, it wouldn't deal with the case where there's two -1ers...

**Comment by blankenberg** [comment]
@natefoo, indeed, if you have 2 -1ers, they could 'protect' each other from removal, unless they can both be removed simultaneously. However, if we do allow multiple committers to be removed simultaneously, could a single rogue committer (or very small group) remove everyone else without recourse?

**Comment by natefoo** [comment]
@blankenberg,

> new members may be added to this group by opening a pull request adding a qualified person to this list

Suggests to me one person at a time.

**Comment by jmchilton** [comment]
Yes this was the unwritten intention, I will clarify that people should be removed one at a time from the committer list.

**Comment by natefoo** [comment]
So, is anyone actually opposed to adding veto override language?

**Comment by jmchilton** [comment]
Pragmatically speaking - I am somewhat opposed to doing that as part of this PR at this point. I think we should merge this as is and then open a PR with a response plan for vetos that we can use this process to respond to.

I don't want a tenuous agreement to fall apart based on conflicts on this issue.
